File: extract-templates.js
/**
 * Template Extraction Script
 *
 * Extracts templates from defaultTemplates.ts to individual JSON files
 */

const fs = require('fs');
const path = require('path');

// Read the defaultTemplates.ts file
const templateFile = fs.readFileSync(
  path.join(__dirname, 'frontend/src/services/templates/defaultTemplates.ts'),
  'utf-8'
);

// Extract the createJinjaTemplate function
const createJinjaTemplateMatch = templateFile.match(
  /function createJinjaTemplate\([\s\S]*?\n\}\)/
);

if (!createJinjaTemplateMatch) {
  console.error('Could not find createJinjaTemplate function');
  process.exit(1);
}

// Extract all template calls
const templateCallRegex = /createJinjaTemplate\(\s*'([^']+)'\s*,\s*'([^']+)'\s*,\s*'([^']+)'\s*,\s*'([^']+)'\s*,\s*`([\s\S]*?)`\s*,\s*(\[[\s\S]*?\])\s*,\s*(\{[\s\S]*?\})\s*,\s*(\[[\s\S]*?\])\s*\)/g;

const templates = [];
let match;

while ((match = templateCallRegex.exec(templateFile)) !== null) {
  const [
    fullMatch,
    id,
    name,
    description,
    category,
    jinjaTemplate,
    variablesStr,
    defaultValuesStr,
    tagsStr
  ] = match;

  try {
    // Parse the variables array
    const variables = eval(variablesStr);

    // Parse the default values object
    const defaultValues = eval(defaultValuesStr);

    // Parse the tags array
    const tags = eval(tagsStr);

    templates.push({
      id,
      name,
      description,
      category,
      template: {
        name,
        description,
        data: {
          label: defaultValues.label || name
        },
        jinjaTemplate,
        variables,
        defaultValues,
        style: {},
        isPublic: true,
        tags
      }
    });
  } catch (error) {
    console.error(`Error parsing template ${id}:`, error.message);
  }
}

// Also extract the custom HTML template (which doesn't use createJinjaTemplate)
const customTemplateMatch = templateFile.match(
  /id: 'template-custom-html'[\s\S]*?\}\s*,\s*\[[\s\S]*?\]\s*\)\s*\]/
);

if (customTemplateMatch) {
  try {
    // This is more complex, we'll handle it manually
    console.log('Found custom HTML template');
  } catch (error) {
    console.error('Error parsing custom template:', error.message);
  }
}

console.log(`Found ${templates.length} templates`);

// Write templates to JSON files
const categories = {
  database: [],
  service: [],
  infrastructure: [],
  external: [],
  component: [],
  container: [],
  custom: [],
  sample: []
};

templates.forEach(template => {
  if (categories[template.category]) {
    categories[template.category].push(template);
  }
});

// Create directory structure
const baseDir = path.join(__dirname, 'frontend/src/services/templates/built-in');

Object.entries(categories).forEach(([category, categoryTemplates]) => {
  const categoryDir = path.join(baseDir, category);

  if (!fs.existsSync(categoryDir)) {
    fs.mkdirSync(categoryDir, { recursive: true });
  }

  categoryTemplates.forEach(template => {
    const filename = `${template.id.replace('template-', '')}.json`;
    const filepath = path.join(categoryDir, filename);

    fs.writeFileSync(
      filepath,
      JSON.stringify(template, null, 2),
      'utf-8'
    );

    console.log(`Created: ${filepath}`);
  });
});

console.log('Extraction complete!');

File: extract_templates.py
"""
Template Extraction Script

Extracts templates from defaultTemplates.ts to individual JSON files
"""

import re
import json
import os
from pathlib import Path

# Read the defaultTemplates.ts file
template_file_path = Path(__file__).parent / 'frontend/src/services/templates/defaultTemplates.ts'

with open(template_file_path, 'r', encoding='utf-8') as f:
    content = f.read()

# Find all createJinjaTemplate calls
pattern = r"createJinjaTemplate\(\s*'([^']+)'\s*,\s*'([^']+)'\s*,\s*'([^']+?)'\s*,\s*'([^']+)'\s*,\s*`([^`]+?)`\s*,\s*(\[[^\]]+?\])\s*,\s*(\{[^}]+?\})\s*,\s*(\[[^\]]+?\])\s*\)"

matches = re.finditer(pattern, content, re.DOTALL)

templates = []

for match in matches:
    id = match.group(1)
    name = match.group(2)
    description = match.group(3)
    category = match.group(4)
    jinja_template = match.group(5).strip()
    variables_str = match.group(6)
    default_values_str = match.group(7)
    tags_str = match.group(8)

    try:
        # Parse variables (convert JavaScript to Python)
        variables_json = variables_str.replace('true', 'True').replace('false', 'False').replace('null', 'None')
        variables = eval(variables_json)

        # Parse default values
        defaults_json = default_values_str.replace('true', 'True').replace('false', 'False').replace('null', 'None')
        default_values = eval(defaults_json)

        # Parse tags
        tags_json = tags_str.replace('true', 'True').replace('false', 'False').replace('null', 'None')
        tags = eval(tags_json)

        template = {
            "id": id,
            "name": name,
            "category": category,
            "template": {
                "name": name,
                "description": description,
                "data": {
                    "label": default_values.get("label", name)
                },
                "jinjaTemplate": jinja_template,
                "variables": variables,
                "defaultValues": default_values,
                "style": {},
                "isPublic": True,
                "tags": tags
            }
        }

        templates.append(template)
        print(f"Parsed template: {id}")

    except Exception as e:
        print(f"Error parsing template {id}: {e}")

# Also try to find the custom HTML template
custom_pattern = r"id:\s*['\"]template-custom-html['\"]"

custom_match = re.search(custom_pattern, content)

if custom_match:
    print("Found custom HTML template (will be handled manually)")

print(f"\nTotal templates extracted: {len(templates)}")

# Create directory structure and save templates
base_dir = Path(__file__).parent / 'frontend/src/services/templates/built-in'

categories = {
    'database': [],
    'service': [],
    'infrastructure': [],
    'external': [],
    'component': [],
    'container': [],
    'custom': [],
    'sample': []
}

# Categorize templates
for template in templates:
    category = template['category']
    if category in categories:
        categories[category].append(template)

# Create files
for category, category_templates in categories.items():
    if not category_templates:
        continue

    category_dir = base_dir / category
    category_dir.mkdir(parents=True, exist_ok=True)

    for template in category_templates:
        # Generate filename from template ID
        filename = template['id'].replace('template-', '') + '.json'
        filepath = category_dir / filename

        # Write JSON file
        with open(filepath, 'w', encoding='utf-8') as f:
            json.dump(template, f, indent=2, ensure_ascii=False)

        print(f"Created: {filepath}")

print("\nExtraction complete!")
